Rebase Workspaces on Editable packages

memsharded opened the pull request:
Changelog: Feature: Re-implement Workspaces based on Editable packages.
Docs: https://github.com/conan-io/docs/pull/1086


Close #3408
Close #3274
Close #3145
Close #3141
Close #3306
Close #4558
Close #4397
Close #4009 
Close #3403
Close #4422
Close #1377

Open issues:

lasote commented:
Is this still a WIP?
(solve conflicts)

memsharded commented:
Ready for review.

danimtb commented on conans/test/integration/workspace_test.py:
```diff
@@ -80,166 +64,255 @@ def package_info(self):
+        client.save(a, path=A)
+
+        project = dedent("""
+                        HelloB/0.1@lasote/stable:
```
Fix indentation

danimtb commented on conans/test/integration/workspace_test.py:
```diff
@@ -80,166 +64,255 @@ def package_info(self):
-        self.assertIn("HelloA/root@project/develop: Applying build-requirement: Tool/0.1@user/testing",
-                      client.out)
+        project = dedent("""
+                            HelloB/0.1@lasote/stable:
```
Fix indentation

danimtb commented on conans/test/integration/workspace_test.py:
```diff
@@ -80,166 +64,255 @@ def package_info(self):
+        client.save(a, path=A)
+
+        project = dedent("""
+                        HelloB/0.1@lasote/stable:
```
Fix indentation

danimtb commented on conans/test/integration/workspace_test.py:
```diff
@@ -80,166 +64,255 @@ def package_info(self):
+                        build/{settings.build_type}/lib
+                        """)
+        metacmake = dedent("""
+                        cmake_minimum_required(VERSION 3.3)
```
Fix indentation

memsharded commented on conans/client/command.py:
```diff
@@ -1388,6 +1388,26 @@ def alias(self, *args):
         self._conan.export_alias(args.reference, args.target)
 
+    def workspace(self, *args):
+        """ handle workspaces
```
This is to be discussed, if we want to go with the ``conan workspace install`` command, or simplify it to ``conan install conanws.yml`` or something like that.

jgsogo commented on conans/client/cache/editable.py:
```diff
@@ -42,3 +42,6 @@ def remove(self, ref):
         return False
+
+    def update(self, workspace_edited):
+        self._edited_refs.update(workspace_edited)
```
_update -vs- substitution._ 

I see no concerns right now, but I wanted to point this out and think about use cases.

jgsogo replied:
Still needed? I saw that you were doing `self._cache.editable_packages = workspace.get_editable_dict()` when creating the workspace.

----
IMO I prefer to keep the information of the editable packages being used inside this class (easier to debug), so I would vote for some code like:

_conan_api.py_
```python

    with self._cache.editable_packages.override(workspace.get_editable_dict()):

         recorder = ActionRecorder()
         deps_graph, _ = self._graph_manager.load_graph(workspace.root, None, graph_info, build,
                                                       False, update, remote_name, recorder)

         print_graph(deps_graph, self._user_io.out)

         installer = BinaryInstaller(self._cache, self._user_io.out, self._remote_manager,
                                    recorder=recorder, hook_manager=self._hook_manager)
         installer.install(deps_graph, keep_build=False, graph_info=graph_info)
         workspace.generate(cwd, deps_graph)
```

_editable.py_
```python
    ...
    @contextmanager
    def override(self, editable_packages):
        old = self._edited_refs
        try:
            self._edited_refs = editable_packages
            yield
        finally:
            self._edited_refs = old
```

memsharded replied:
Fixing it. I won't use the context manager, because actually, the @api_method should manage to provide a clean environment of ``self._cache`` (already opened engineering issue)

memsharded commented on conans/client/conan_api.py:
```diff
@@ -456,6 +458,35 @@ def download(self, reference, remote_name=None, package=None, recipe=False):
+        if os.path.isabs(path):
+            abs_path = path
+        else:
+            abs_path = os.path.normpath(os.path.join(cwd, path))
```
Why? You need to check if it is conditional? I don't understand.

jgsogo replied:
Yes, the `abs_path = os.path.normpath(os.path.join(cwd, path))` will work for already absolute paths.

danimtb replied:
this ``os.path.normpath(os.path.join(cwd, path))`` is the same as using directly ``os.path.abspath(path)``. Wouldn't that work even if ``path`` is already absolute?

memsharded replied:
Not the same, as long as ``cwd`` is an argument to the api.

danimtb replied:
Since you are creating a new api method, is ``cwd`` parameter really needed?

memsharded replied:
Yes, as all the other api methods, that already have ``cwd`` argument. 
We need to keep consistency. We could discuss over the conan_api as a whole somewhere else, but not here.

danimtb replied:
There are some api methods that have a path parameter that do not have the ``cwd`` one. I don't see the need of having such a parameter in the api and here is not necessary either.

We could move the discussion to an engineering issue but I am still reluctant to include a new api method with a "useless" parameter...

memsharded replied:
There are some API methods that the cwd really means something (for example, where to output files), and other that is not used at all, and only some path is needed. 

It is not useless, it means that the ``conanworkspace.cmake`` will be generated in that ``cwd`` if you use the python API, otherwise you need to do a chdir before the call (and undo it later, which is ugly and error prone). The ``cwd`` stays, until we decide to remove it from the whole api.

jgsogo commented on conans/client/installer.py:
```diff
@@ -329,6 +320,25 @@ def _handle_node_editable(self, node):
 
+            build_folder = editable_cpp_info.folder(node.ref, "build",
+                                                    settings=node.conanfile.settings,
+                                                    options=node.conanfile.options)
```
`editable_cpp_info` object is created from a `package_layout`, so it shouldn't accept a `ref` in its functions, it must be already coupled with a reference.

memsharded replied:
I didn't put this here, it is just copied form the above ``editable_cpp_info.apply_to(node.ref,``
I agree this should be fixed, but that is another PR, this is focused on the Workspaces

jgsogo replied:
Didn't realize about the previous, I thought it was the first time, ok, you win 😆

jgsogo commented on conans/client/installer.py:
```diff
@@ -329,6 +320,25 @@ def _handle_node_editable(self, node):
+                                                    settings=node.conanfile.settings,
+                                                    options=node.conanfile.options)
+            if build_folder is not None:
+                build_folder = os.path.join(package_layout.conan(), build_folder)
```
Use `base_path` instead of calling again to `package_layout.conan()`

jgsogo commented on conans/model/editable_cpp_info.py:
```diff
@@ -44,22 +56,26 @@ def load(filepath):
             ref, cpp_info_dir = section.rsplit(":", 1) if ':' in section else (None, section)
+            if cpp_info_dir == "folders":
+                folders[ref] = {k: v for k, v in parser.items(section)}
+                continue
```
This `folders` section breaks the previous layout of this file. It might be better to keep those folders under control (we can rename the `cpp_info_dirs` or create and additional one):

```ini
...

[<ref>:build]
/path/...
.../path
```

there is no need to:

```ini
[<ref>:folders]
build=path/at
```

memsharded replied:
The problem is that those folders are unique, not a list of folders. Using the section approach it is very easy to fall in that error, you did in your example.

jgsogo replied:
We can check there is only one path here

jgsogo commented on conans/model/editable_cpp_info.py:
```diff
@@ -44,22 +56,26 @@ def load(filepath):
                                      % (cpp_info_dir, filepath))
             if ref:
                 try:
                     r = ConanFileReference.loads(ref)
```
It was missing before, but add `validate=True` in the constructor (remember that the default value should be changed to `False` at some moment in the future)

jgsogo commented on conans/test/integration/workspace_test.py:
```diff
@@ -80,166 +64,255 @@ def package_info(self):
-"""
-        client.save({WORKSPACE_FILE: project})
-        client.run("install . -if=build")
+        client.save(a, path=A)
```
```suggestion
        client.save(a, path=os.path.join(client.current_folder, "A"))
```

jgsogo commented on conans/test/integration/workspace_test.py:
```diff
@@ -80,166 +64,255 @@ def package_info(self):
+        client.runner(cmd_debug, cwd=client.current_folder)
+        self.assertIn("Hello World C Debug!", client.out)
+        self.assertIn("Hello World B Debug!", client.out)
+        self.assertIn("Hello World A Debug!", client.out)
```
Parameterize this test for `build_type` in `Debug|Release`

memsharded replied:
No, I don't want to do that. I want to run one build after the other. That is the key of this test, in fact that helped to solve many bugs. Behavior is really affected by dealing with both builds.

jgsogo commented on conans/test/integration/workspace_test.py:
```diff
@@ -80,166 +64,255 @@ def package_info(self):
+        a["src/CMakeLists.txt"] += ("add_executable(app main.cpp)\n"
+                                    "target_link_libraries(app helloA)\n")
+        a["src/main.cpp"] = main_cpp
+        client.save(a, path=A)
```
```suggestion
        client.save(a, path=os.path.join(client.current_folder, "A"))
```

jgsogo commented on conans/test/integration/workspace_test.py:
```diff
@@ -80,166 +64,255 @@ def package_info(self):
+        base_debug = os.path.join(client.current_folder, "build_debug")
+
+        with client.chdir("build_release"):
+            client.run("workspace install ../conanws.yml")
```
Can we run the first part of the test inside this context and the second one inside the `build_debug` one? Also, why not parameterize?

memsharded replied:
I have tried, but there are some lines in between that doesn't run in that folder, and makes it a bit cluttered with a lot of ``with client.chdir``. So leaving this as is.

Not parameterize, because the important thing here is that they run in the same execution, both debug and release might interfere with each other. They did, I solved a lot of bugs because the test runs both in the same execution.

jgsogo commented on conans/test/integration/workspace_test.py:
```diff
@@ -249,81 +322,159 @@ def files(name, depend=None):
+        a["src/CMakeLists.txt"] += ("add_executable(app main.cpp)\n"
+                                    "target_link_libraries(app helloA)\n")
         a["src/main.cpp"] = main_cpp
         client.save(a, path=A)
```
```suggestion
        client.save(a, path=os.path.join(client.current_folder, "A"))
```

jgsogo commented on conans/test/integration/workspace_test.py:
```diff
@@ -249,81 +322,159 @@ def files(name, depend=None):
+
+        self.assertNotIn("Release", client.out)
+
+    def build_requires_test(self):
```
I thought this test was going to use a `build_requires` in editable mode. Does it worth another test?

memsharded replied:
Yes, added.

memsharded commented:
Reviewed:

- Converted command link=>editable
- Search is able to show editable packages too
- Changed [folders] => [build_folder][source_folder] in layout files
- Definition of workspace invalidates other editable packages

jgsogo commented on conans/client/command.py:
```diff
@@ -1388,50 +1388,72 @@ def alias(self, *args):
+        subparsers = parser.add_subparsers(dest='subcommand', help='sub-command help')
+
+        install_parser = subparsers.add_parser('install', help='install this workspace')
+        install_parser.add_argument('path', help='path to workspace file or folder')
```
Add something related to the default name of the _workspace file_, was it `conan_ws.yml`?

memsharded replied:
There is no default name now.

jgsogo replied:
Ok, then it can no longer be a _path to workspace folder_

jgsogo commented on conans/client/conan_api.py:
```diff
@@ -456,6 +460,31 @@ def download(self, reference, remote_name=None, package=None, recipe=False):
+        workspace = Workspace(abs_path, self._cache)
+        graph_info = get_graph_info(profile_name, settings, options, env, cwd, None,
+                                    self._cache, self._user_io.out)
+        self._user_io.out.success("Using workspace file from %s" % workspace._base_folder)
```
Instead of the `base_folder` we should inform here of the file actually used (and better not to use a private member).

memsharded replied:
Removed, this doesn't make sense anymore that it will fail if the file is not found.

jgsogo commented on conans/model/editable_cpp_info.py:
```diff
@@ -44,22 +57,30 @@ def load(filepath):
         except configparser.Error as e:
             raise ConanException("Error parsing layout file: %s\n%s" % (filepath, str(e)))
         data = OrderedDict()
+        folders = {}
```
We don't need a new member to store these _folders_, they can go just into the _data_ one.

memsharded replied:
No, better not mix it:

```python
if data:  # Invalidate previously existing dirs
            for info_dir in self.cpp_info_dirs:
                setattr(cpp_info, info_dir, [])
```
The definition of a ``build_folder`` would invalidate includedirs, etc, and I think this is not desired.

jgsogo replied:
I would double check this. This is against the principle we established when talking about the editable layouts in the whiteboard: only one source of truth. IMO it is ok to invalidate the `cpp_info dirs` also when it is only the `build_folder` (or the source one) in the layout file.

memsharded replied:
Not nice, because you cannot define those folders in the ``package_info()`` and ``cpp_info``. So you can have a perfectly valid, customized ``package_info()``, but you still need to define your ``build_folder`` in the layout, and certainly you don't want to invalidate the ``cpp_info``. 
One thing is layout (build_folder, source_folder), and the other is cpp_info.

jgsogo replied:
I understand your concern, but still I think that two sources of truth can be confusing. With the _double check_ I wanted to share this consideration with the team, to see what they think about it. 

Please, @conan-io/barbarians , your opinion on this thread?

memsharded replied:
It is not that it would be using 2 sources of truth. Not allowing this will forbid the possibility to use the information from ``package_info()`` and ``cpp_info``, as the ``build_folder`` is necessary for workspaces, and that cannot be defined in ``package_info()``.

lasote replied:
I agree @memsharded. 
Shouldn't this module be renamed too? Do you know if doing "git rename" the diff is smart enough?

memsharded replied:
Yeah, I didn't want to rename it yet (same with WorkspaceTest), because when there are % of changes, then it detects as a new file, and it is impossible to review. I will do those renames and moves after, in another PR.

jgsogo commented on conans/search/search.py:
```diff
@@ -99,16 +99,12 @@ def search_recipes(cache, pattern=None, ignorecase=True):
-        return sorted(ret)
+    refs = [ConanFileReference(*folder.split("/")) for folder in subdirs]
+    refs.extend(cache.editable_packages.refs().keys())
+    refs = sorted(refs)
```
Order the list after applying the filter (scratch a little performance)

memsharded replied:
True, done

lasote commented:
Just remember to add the docs before merge when it's ready.

jgsogo commented on conans/client/conan_api.py:
```diff
@@ -457,6 +462,40 @@ def download(self, reference, remote_name=None, package=None, recipe=False):
+            if node.recipe == RECIPE_EDITABLE:
+                generators = workspace[node.ref].generators
+                if generators is not None:
+                    node.conanfile.generators = generators
```
Do we need to override? Can we append to existing ones? There could be recipes using custom generators (like Qt does) to create needed files.

jgsogo commented on conans/client/cache/editable.py:
```diff
@@ -21,6 +21,9 @@ def __init__(self, cache_folder):
         else:
             self._edited_refs = {}  # {ref: {"path": path, "layout": layout}}
 
+    def refs(self):
```
For naming conventions: this is a dictionary containing `refs` as keys, but not a list of `refs`... but we can postpone this renaming to a future PR dedicated to it.

jgsogo commented on conans/client/installer.py:
```diff
@@ -327,6 +318,25 @@ def _handle_node_editable(self, node):
                                        settings=node.conanfile.settings,
                                        options=node.conanfile.options)
 
+            build_folder = editable_cpp_info.folder(node.ref, "build_folder",
```
Is it worth to add constants for `build_folder` adn `source_folder` in the `EditableLayout` class?

jgsogo commented on conans/client/installer.py:
```diff
@@ -327,6 +318,25 @@ def _handle_node_editable(self, node):
+                output.info("Generated %s" % BUILD_INFO)
+                # Build step might need DLLs, binaries as protoc to generate source files
+                # So execute imports() before build, storing the list of copied_files
+                from conans.client.importer import run_imports
```
Is it needed to hide the import here?

jgsogo commented on conans/model/workspace.py:
```diff
@@ -1,155 +1,121 @@
-        if self._install_folder:
-            pkg_folder = os.path.join(self._install_folder, self._conanfile_folder, folder)
+        self._conanfile_folder = data.get("path")  # The folder with the conanfile
+        layout = data.get("layout")
```
All these lines trying to resolve the layout file can be done before entering the constructor, in the `Workspace` object that is creating the `LocalPackage`s.

Also, I would unroll the members of the `data` object so this class doesn't need to know about the workspaces definition file layout.

memsharded replied:
Leaving this for later

jgsogo commented on conans/model/workspace.py:
```diff
@@ -1,155 +1,121 @@
-                raise ConanException("Can't generate CMakeLists.txt, will overwrite existing one")
+
+    def generate(self, cwd, graph):
+        editables = {node.ref: node.conanfile for node in graph.nodes
```
No need for this dictionary, it is only used to retrieve the `conanfile` in line 52 where you can use `node.conanfile`.

jgsogo replied:
😿

memsharded replied:
Fixed

jgsogo replied:
😸

jgsogo commented on conans/model/workspace.py:
```diff
@@ -1,155 +1,121 @@
-        self._generator = None
-        self._name = "ConanWorkspace"
+    def __init__(self, path, cache):
+        self._cache = cache
```
I think we can avoid passing the `cache` object here. It is used just in the `Workspace::generate` function (ok, also to get the `cache.conan_folder`, but it can be easily substituted) from the `conan_api` where we already know which packages are in _editable mode_, so we can get there the `package_layouts` we are going to use and pass them to this class.

memsharded replied:
leaving this for later

lasote commented on conans/model/workspace.py:
```diff
@@ -116,12 +121,18 @@ def _loads(self, text):
-        editables = yml.pop("editables", None)
+        editables = yml.pop("editables", {})
         for package_name, data in editables.items():
+            if not data:
```
Shouldn't the validation go to the `LocalPackage`? Which fields are mandatory? Only the `path`?

memsharded replied:
Yes, only the path is really mandatory

lasote replied:
Then it should complain because the path is missing, the message about the missing data is a bit weird.

memsharded replied:
Done

lasote commented on conans/model/workspace.py:
```diff
@@ -12,11 +12,13 @@
 class LocalPackage(object):
-    def __init__(self, base_folder, data, cache, ws_layout, ws_generators):
+    def __init__(self, base_folder, data, cache, ws_layout, ws_generators, ref):
+        if not data:
```
Could we move here both checks to avoid the duplicated raise?
` if not data or not data.get("path"):` ?

memsharded replied:
Not that easy, because I am not doing a ``data.get()``, but a ``data.pop()`` to extract the field and that cannot be done in a single line.

lasote replied:
Mhh I'm missing something probably, but if you first check if the data["get"] is there or not, the following check after the data.pop("get") is not necessary, right?

memsharded replied:
Ok, right, done.
